### PR TITLE
Allow writers to overwrite existing data

### PR DESCRIPTION
**Pull request by JAEarly**
## Description of changes:

CloudUploader has an `exist_ok` kwarg that is set to False by default.  
Writers (e.g. MDSWriter) cannot overwrite this.  

This PR makes two changes:
1. `exist_ok` can now be set by a writer.
2. If `exist_ok` is True, existing files are now deleted. This allows easy debugging when developing streaming datasets as the files do not have to be manually deleted each time the code is run.

Tests and docs updated accordingly.

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [x] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/storage/upload.py [hunk 1]
```diff
@@ -171,8 +171,9 @@ def __init__(self,
                 raise FileExistsError(f'Directory is not empty: {self.local}')
```
Comment (karan6181) [inline streaming/base/storage/upload.py:171]: Can you please extend the error message saying that you set `exist_ok=True` if you would like to delete the content from the local directory and start fresh?

File: streaming/base/format/base/writer.py [hunk 2]
```diff
@@ -100,7 +100,8 @@ def __init__(self,
 
         # Validate keyword arguments
         invalid_kwargs = [
-            arg for arg in kwargs.keys() if arg not in ('progress_bar', 'max_workers', 'retry')
+            arg for arg in kwargs.keys()
+            if arg not in ('progress_bar', 'max_workers', 'retry', 'exist_ok')
```
Comment (karan6181) [inline streaming/base/format/base/writer.py:104]: Can you please add the args comment for `exist_ok` at the start of the method? For example, something like below or in your own words
```
exist_ok (bool): If the local directory exists and is not empty, whether to overwrite
                the content or raise an error. `False` raises an error. `True` deletes the
                content and starts fresh. Defaults to `False`.
```

Also, do the same for all `writer.py` file in directory https://github.com/mosaicml/streaming/tree/main/streaming/base/format.

**Review by karan6181 — COMMENTED** [review]
Thank you for creating the PR. The changes looks good to me. I have posted some minor comments.

@XiaohanZhangCMU Can you please also take a look?

File: streaming/base/storage/upload.py (hunk 1 above)
Comment (JAEarly) [reply streaming/base/storage/upload.py]: Thanks for the quick response! Done.

**Comment by JAEarly** [comment]
Looking at the tests that are failing: I believe there's now a conflict when `exists_ok=True` and `keep_local=True`, or at least this seems to be what the GitHub test logs says. I'm having issues running the failing tests locally so I'm unable to debug. @karan6181 do you have further insight on this?

**Comment by XiaohanZhangCMU** [comment]
@JAEarly Currently exists_ok=True in clouduploader.py is only used by merging index (streaming/base/utils/merge_index). Your change will make that fail. 

I suggest you add the "file-removing behavior" to mdswriter instead of upload.py.

File: streaming/base/storage/upload.py [hunk 3]
```diff
@@ -168,11 +168,14 @@ def __init__(self,
...
-                raise FileExistsError(f'Directory is not empty: {self.local}')
+                raise FileExistsError(
+                    f'{self.local} not empty. Set exist_ok to True to overwrite the current contents (will delete).'
+                )
             else:
                 logger.warning(
-                    f'Directory {self.local} exists and not empty. But continue to mkdir since exist_ok is set to be True.'
+                    f'Directory {self.local} exists and is not empty; exist_ok is set to True so will remove contents.'
                 )
+                shutil.rmtree(self.local)
```
Comment (XiaohanZhangCMU) [inline streaming/base/storage/upload.py:178]: This will break all the other code that does not perform writing.

**Review by XiaohanZhangCMU — CHANGES_REQUESTED** [review]

File: streaming/base/storage/upload.py (hunk 3 above)
Comment (JAEarly) [reply streaming/base/storage/upload.py:178]: Thank you for the input, I'll change the implementation now.

Comment (JAEarly) [reply streaming/base/storage/upload.py]: Okay, I've moved the removal code into writer which should fix the issue.

**Review by XiaohanZhangCMU — APPROVED** [review]
Lgtm. Wait for ci checks /

**Comment by JAEarly** [comment]
Should be good to merge, I've rebased just need the CI checks to be run again, thanks!

File: streaming/base/format/base/writer.py [hunk 4]
```diff
@@ -116,6 +120,13 @@ def __init__(self,
 
         self.shards = []
 
+        # Remove local directory if requested prior to creating writer
+        local = out if isinstance(out, str) else out[0]
+        if os.path.exists(local) and kwargs.get('exist_ok', False):
```
Comment (XiaohanZhangCMU) [inline streaming/base/format/base/writer.py:125]: shouldn't this be kwargs.get('exist_ok', True)?

**Review by XiaohanZhangCMU — CHANGES_REQUESTED** [review]

File: streaming/base/format/base/writer.py (hunk 4 above)
Comment (JAEarly) [reply streaming/base/format/base/writer.py:125]: I would say the directory should not be removed by default - files should not be deleted unless the user has explicitly set `exist_ok` to True.

Comment (JAEarly) [reply streaming/base/format/base/writer.py:125]: My understanding is:
- If `local` exists but `exist_ok` is False (default value), then `local` will not be removed and `CloudUploader` will throw an error. Avoids accidentally overwriting data.
- If `local` doesn't exist then nothing needs deleting and `CloudUploader` will work fine.
- If `local` exists and `exist_ok` is True (has to be set by the user), then `local` will be removed with a warning and `CloudUploader` will work fine.

**Review by XiaohanZhangCMU — APPROVED** [review]
LGTM! Thanks for the improvement!

**Pull request merged**